core/src/config.rs

```diff
@@ -85,51 +82,70 @@ impl Default for NetworkConfiguration {
 }
 
 impl Configuration {
-    /// This method will build `Configuration` from a json *pretty* formatted file (without `:` in
-    /// key names).
+    /// Construct [`Self`] from a path-like object.
     ///
-    /// # Panics
-    /// This method will panic if configuration file presented, but has incorrect scheme or format.
+    /// The method has automatic de-duplication for items that are
+    /// represented as [`HashSet`]s, and errors if a duplicate is
+    /// found for [`crate::sumeragi::config::TrustedPeers`], i.e. if
+    /// two peers have the same public key.
     ///
     /// # Errors
-    /// This method will return error if system will fail to find a file or read it's content.
-    pub fn from_path<P: AsRef<Path> + Debug>(path: P) -> Result<Configuration> {
-        let file = File::open(path).wrap_err("Failed to open the config file")?;
+    /// - File not found.
+    /// - File found, but peer configuration parsing failed.
+    /// - File re-opening failed.
+    /// - Length of the array in raw json representation is different
+    /// to the lenght of the array in
+    /// [`self.sumeragi.trusted_peers.peers`], most likely due to two
+    /// (or more) peers having the same public key.
+    pub fn from_path<P: AsRef<Path> + Debug + Clone>(path: P) -> Result<Configuration> {
+        let file = File::open(path.clone())
+            .wrap_err(format!("Failed to open the config file {:?}", path))?;
         let reader = BufReader::new(file);
-        let mut configuration: Configuration =
-            serde_json::from_reader(reader).wrap_err("Failed to deserialize json from reader")?;
-        configuration.sumeragi.key_pair = KeyPair {
-            public_key: configuration.public_key.clone(),
-            private_key: configuration.private_key.clone(),
-        };
-        configuration.sumeragi.peer_id =
-            PeerId::new(&configuration.torii.p2p_addr, &configuration.public_key);
+        let mut configuration: Configuration = serde_json::from_reader(reader).wrap_err(
+            format!("Failed to parse {:?} as Iroha peer configuration.", path),
+        )?;
+        // TODO: While re-reading the file isn't too much of a problem,
+        // from_reader drops the file, it's safe to re-read.
+        let second_reader = BufReader::new(File::open(path.clone())?);
+        let raw_json: serde_json::Value = serde_json::from_reader(second_reader).wrap_err(format!("Failed to read raw json from {:?}. However, parsing the file as a Configuration structure succeeded. ", path))?;
+        if let serde_json::Value::Array(arr) = &raw_json["SUMERAGI"]["TRUSTED_PEERS"] {
+            if arr.len() != configuration.sumeragi.trusted_peers.peers.len() {
```

Comment:
I'm a bit weary of this approach. What if `TrustedPeers` was changed to be represented with `Vec` instead of `HashSet`?
If you were to add a failing test for this I'd be convinced